doc/developer/design/20231027_refresh_mvs.md

```diff
@@ -0,0 +1,276 @@
+
+### Implementation Alternative: Creating a New Dataflow at Each Refresh
+
+(The above alternatives are alternatives to `REFRESH EVERY` MVs, but this alternative is just about the implementation details of `REFRESH EVERY` MVs.)
+
+This approach would need similar replica management as the proposed solution, but instead of rehydrating an existing dataflow, we would always create a new dataflow to run the refresh. An advantage of this approach is that these dataflows could be single-time dataflows, and thus rely on one-shot SELECT optimizations. However, this approach would be more complicated, as we would have to separate the lifecycle of dataflows from the lifecycle of MVs. These new dataflows would be a kind of hybrid between Peek dataflows and MV dataflows, and would require new code in various parts of the Compute Controller. As we are under some time pressure to implement this feature, we opted for the simpler solution proposed in this document.
+
+### Implementation Alternative: Suspending the Replica Process to Disk
+
+As mentioned in the [scoping section](#out-of-scope), an alternative implementation that would cover also intermediate refresh intervals would be to suspend the replica process and save its entire state to S3 and a local disk cache, and restore it when a refresh is needed. [There are existing techniques for saving process state, e.g., CRIU.](https://faun.pub/kubernetes-checkpointing-a-definitive-guide-33dd1a0310f6) However, this would be a much bigger project than the proposed solution.
+
+## Automated Cluster Scheduling for `REFRESH EVERY` MVs
+
+TODO: updated discussion
+
+## Introspection / Observability
+
+For showing the `REFRESH` options of each materialized view, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refresh_options`, which will have a row for each non-trivial refresh option of each MV. The columns will be `(materialized_view_id text, interval interval, interval_aligned_to timestamptz, at timestamptz)`, and only those columns will have a non-null value that are relevant for the refresh option that the row is about (e.g., a `REFRESH EVERY` would be non-null in interval and interval_aligned_to). `REFRESH ON COMMIT` is considered a trivial refresh option, and won't show up in the table. (See Nikhil's [msg](https://materializeinc.slack.com/archives/C06535JL58R/p1701228087745239?thread_ts=1700595291.424189&cid=C06535JL58R) on an old design thread.)
+
+For showing the previous and next refresh of each MV, I'm planning to create a new table in `mz_internal` called `mz_materialized_view_refreshes (materialized_view_id text, last_completed_refresh timestamptz, next_refresh timestamptz)`. (The values would be derived from the write frontier of the MV's storage collection.)
```

Comment:
What is the mechanism responsible for updating this table?

Reply:
I'm thinking to update it in `Controller::record_frontiers`, where `mz_frontiers` is also updated. Would that make sense?

The values will be calculated as follows:
- `next_refresh` will usually be the write frontier of the MV's storage collection, unless:
  - The write frontier is 0. This happens when the first refresh hasn't completed yet, in which case we need to figure out the first refresh: we'll call `RefreshSchedule::round_up_timestamp` on the MV's `as_of`.
  - The write frontier is `[]`. This happens when the MV is after its last refresh, and will never be refreshed again. We could represent this simply with a `NULL`.
- `last_completed_refresh` is roughly `RefreshSchedule::round_down_timestamp(next_refresh - 1)`, plus special handling for the `[]` write frontier case. (`round_down_timestamp` will be  a new function, which will be the same as `round_up_timestamp`, but backwards.)

Reply:
I think `Controller::record_frontiers` is not the best place! That method exists only because we want to record frontiers of both compute and storage collections, so there needs to be some code that coordinates (i.e. queries the frontiers from compute and reports them to storage for enriching and writing down). MV refreshes are a pure compute concept and so I think they should be handled internally to the compute controller.

We already have a couple compute-only introspection collections and you can check how they are written. The exact details differ but they all have in common that some part of the controller state writes to an `introspection_tx` and then the top-level compute controller collection updates from the corresponding `introspection_rx` and writes them to persist. The way I imagine this would work for `REFRESH` MVs probably similar to the hydration status logging. When a new `FrontierUpper` for a `REFRESH` MV arrives, it is passed to some refresh tracking logic that emits new updates into the `introspection_tx`. Probably you'd also want to have a wrapping type like `HydrationState` to make sure any emitted rows are retracted when the MV is dropped.

Reply:
Thanks, updated the design doc.